### DISCUSSION TITLE
add getDefaults method

### DISCUSSION THREAD

**Discussion by NurbekGithub** [47ng/nuqs#584] [Ideas] [answered]
Valibot and Zod has a getDefaults method where you can get default values from schema.
It would be useful in cases when you want to use those values somewhere else, e.g. to check if filters' state dirtiness.

**Comment by franky47** [comment]
You can get the default value from a configured parser:

```ts
import { parseAsInteger } from 'nuqs'

const parser = parseAsInteger.withDefault(42)

parser.defaultValue // 42
```

**Reply by NurbekGithub** [reply]
oh thanks, but it would be great if we can get default values from schema itself, instead of manually building default values object. wdyt?

**Reply by franky47** [reply]
I'm not sure I follow. Do you have an example of the API you have in mind?

**Reply by NurbekGithub** [reply]
sure something like
```js
import {getDefaultValues} from 'nuqs'
const searchSchema = {
	page: parseAsInteger.withDefault(0),
	search: parseAsString.withDefault(''),
	status: parseAsStringLiteral(['a', 'b']),
};

const defaultValues = getDefaultValues(searchSchema) // = {page: 0, search: '', status: undefined}
```
and you can use the defaultValues as you need. e.g.
```js
const serializer = createSerializer(searchParamsSchema);

const [searchParams, setSearchParams] = useQueryStates(searchParamsSchema);
const isSearchDirty = const isSearchDirty = useMemo(
  () => serializer(searchParams) !== serializer(defaultValues),
  [searchParams]
);
```

**Reply by franky47** [reply, marked as answer]
Your first code example can be done with a simple function (simple at runtime, but harder to make type-safe; PR #578 might help here):

```ts
function getDefaultValues(schema: Record<string, ParserBuilder<any>) {
  const result = {}
  for (const [key, parser] of Object.entries(schema)) {
    result[key] = parser.defaultValue
  }
  return result
}

// Or if you prefer a functional programming approach:
function getDefaultValues(schema: Record<string, ParserBuilder<any>) {
  return Object.entries(schema)
    .reduce((acc, [key, parser]) => ({
      ...acc,
      [key]: parser.defaultValue
    }), {})
}
```

Now, your second code example could work, but going through the serializer could be expensive.

Instead, you can use the `eq` operator in a parser to check against the default value (its purpose is to perform shallow comparison on arrays, objects and other potentially referentially unstable values). And you can access the default value in the same way, avoiding the helper function above altogether:

```ts
const searchParamsSchema = {
  // ...
}


function useIsSearchDirty() {
  const [searchParams] = useQueryStates(searchParamsSchema)
  const isSearchDirty = useMemo(
    () => {
      let dirty = false
      for (const [key, parser] of Object.entries(searchParamsSchema)) {
        const equals = parser.eq ?? Object.is
        if (!equals(searchParams[key], parser.defaultValue)) {
          dirty = true
          break
        }
      }
      return dirty
    },
    [searchParams]
  )
  return isSearchDirty
}
```

**Reply by NurbekGithub** [reply]
thanks a lot for the reply, and for the great work with nuqs overall! really enjoying it
